### ISSUE TITLE
clear_cache query of analysis engine does not free memory

### ISSUE THREAD

**Issue by kevinsung** [lightvector/KataGo#592]
### Steps to reproduce
1. Run the analysis engine as described at https://github.com/lightvector/KataGo/blob/master/docs/Analysis_Engine.md
2. Send enough queries to cause RAM usage to increase noticeably (in my case up to 30GB)
3. Send the `clear_cache` query
4. Get a response which is just the query echoed back, as expected

### Result
RAM usage does not decrease. htop shows the same RAM usage before and after the `clear_cache` query

### Expected
RAM usage should decrease

### Other info
I have tested this on both the stable (version 1.10.0) and master branches. I'm running Arch Linux.

**Comment by lightvector** [comment]
Thanks, I'll investigate.

**Comment by lightvector** [comment]
Sorry for the long delay! I've experimented with this a bit now and done some research, and I think that there isn't much to do about this behavior. What's going on is that depending on operating system and the memory allocator being used for C++, even when memory is freed the memory may not actually be returned to the operating system during the lifetime of the process.

One reason for this is memory fragmentation - the OS may hand memory to a process in pages of 4kb or some other chunky size, and even if almost all the neural net cache is freed, if there is still *anything* allocated on a page, even only some small few-byte object, the entire page still can't be returned to the OS. The other reason is that a lot of memory allocators in C++ don't bother to implement the additional complexity and performance cost of trying to manage the return of pages to the OS, and simply just never return pages.

I tested this and it seem pretty clear what is going on. For example on my test machine I tried a 1-million visit search on a few successive positions with a large maximum cache size, and when clear_cache was run in between each one, the memory usage first went to 2.6G, then next 3.1G (presumably due to a little bit of memory fragmentation where the freed memory space from clear_cache couldn't perfectly be re-used for the next 1 million visits), but then stayed at 3.1G after.

Whereas when running it without clear_cache in between, I observed 2.6G, then 3.8G, then 4.6G.

So, I think this is just what we have to work with, because this is just how C++ works. If you call clear_cache regularly, memory will be freed and re-used instead of continuing to grow much, but whatever the largest usage is will continue to be owned by the process since it won't be returned to the OS. If you do want it to be returned to the OS, then you will need to look up among alternative C++ memory allocators besides the standard glibc malloc (e.g. tcmalloc, jemalloc) and find one of them that does implement return of memory to the OS, and compile KataGo using one of those. Or alternatively, periodically kill and restart the process as a whole to reset the memory usage of the process.
